addons/mail/models/mail_thread.py

```diff
@@ -3648,25 +3648,35 @@ def _notify_thread_by_web_push(self, message, recipients_data, msg_vals=False, *
         """
         msg_vals = dict(msg_vals or {})
         partner_ids = self._extract_partner_ids_for_notifications(message, msg_vals, recipients_data)
-        if not partner_ids:
-            return
-
-        partner_devices_sudo = self.env['mail.push.device'].sudo()
-        devices = partner_devices_sudo.search([
-            ('partner_id', 'in', partner_ids)
-        ])
+        devices, private_key, public_key = self._get_web_push_parameters(partner_ids)
         if not devices:
             return
+        payload = self._truncate_payload(self._notify_by_web_push_prepare_payload(message, msg_vals=msg_vals))
+        self._push_web_notification(payload, devices, private_key, public_key)
 
+    def _get_web_push_parameters(self, partner_ids):
+        """
+        :param partner_ids: IDs of the res.partners
+        :returns: the `mail.push.device` records, the vapid private key and the vapid public key
+        """
+        if not partner_ids:
+            return self.env['mail.push.device'].sudo(), None, None
         ir_parameter_sudo = self.env['ir.config_parameter'].sudo()
         vapid_private_key = ir_parameter_sudo.get_param('mail.web_push_vapid_private_key')
         vapid_public_key = ir_parameter_sudo.get_param('mail.web_push_vapid_public_key')
         if not vapid_private_key or not vapid_public_key:
-            _logger.warning("Missing web push vapid keys !")
-            return
+            return self.env['mail.push.device'].sudo(), None, None
+        partner_devices_sudo = self.env['mail.push.device'].sudo()
+        devices = partner_devices_sudo.search([
+            ('partner_id', 'in', partner_ids)
+        ])
```

Comment:
Can't this be single line?

Also apply formatter/double quotes in whole diff.

Reply:
Double quote is never a guideline in python.